server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+
+            if (!nodeType.equals(druidNode.getNodeType())) {
+              log.error(
+                  "Node[%s:%s] add is discovered by node watcher of nodeType [%s]. Ignored.",
+                  druidNode.getNodeType(),
+                  druidNode,
+                  nodeType
+              );
+              return;
+            }
+
+            log.info("Received event [%s] for Node[%s:%s].", event.getType(), druidNode.getNodeType(), druidNode);
+
+            addNode(druidNode);
+
+            break;
+          }
+          case CHILD_REMOVED: {
+            DiscoveryDruidNode druidNode = jsonMapper.readValue(event.getData().getData(), DiscoveryDruidNode.class);
+
+            if (!nodeType.equals(druidNode.getNodeType())) {
+              log.error(
+                  "Node[%s:%s] removal is discovered by node watcher of nodeType [%s]. Ignored.",
+                  druidNode.getNodeType(),
+                  druidNode,
+                  nodeType
+              );
+              return;
+            }
+
+            log.info("Node[%s:%s] disappeared.", druidNode.getNodeType(), druidNode);
+
+            removeNode(druidNode);
+
+            break;
+          }
+          default: {
+            log.error("Ignored event type [%s] for nodeType [%s] watcher.", event.getType(), nodeType);
+          }
+        }
+      }
+      catch (Exception ex) {
+        log.error(ex, "unknown error in node watcher for type [%s].", nodeType);
+      }
+    }
+
+    private synchronized void addNode(DiscoveryDruidNode druidNode)
+    {
+      DiscoveryDruidNode prev = nodes.put(druidNode.getDruidNode().getHostAndPortToUse(), druidNode);
+      if (prev == null) {
+        for (DruidNodeDiscovery.Listener l : nodeListeners) {
+          listenerExecutor.submit(() -> {
+            try {
+              l.nodeAdded(druidNode);
+            }
+            catch (Exception ex) {
+              log.error(ex, "Exception occured in DiscoveryDruidNode.nodeAdded(node=[%s]) in listener [%s].", druidNode, l);
+            }
+          });
+        }
+      } else {
+        log.warn("Node[%s] discovered but existed already [%s].", druidNode, prev);
+      }
+    }
+
+    private synchronized void removeNode(DiscoveryDruidNode druidNode)
+    {
+      DiscoveryDruidNode prev = nodes.remove(druidNode.getDruidNode().getHostAndPortToUse());
+
+      if (prev == null) {
+        log.warn("Noticed disappearance of unknown druid node [%s:%s].", druidNode.getNodeType(), druidNode);
+        return;
+      }
+
+      for (DruidNodeDiscovery.Listener l : nodeListeners) {
+        listenerExecutor.submit(() -> {
+          try {
+            l.nodeRemoved(druidNode);
+          }
+          catch (Exception ex) {
+            log.error(ex, "Exception occured in DiscoveryDruidNode.nodeRemoved(node=[%s]) in listener [%s].", druidNode, l);
+          }
+        });
+      }
+    }
+
+    public void start()
+    {
+      try {
+        cache.getListenable().addListener(this);
```

Comment:
I think it might be nicer to register an anonymous inner class here instead of having `this` expose the `PathChildrenCacheListener` interface.

Once someone gets one of these objects from the getter above, should they be able to register it with Curator?